### DISCUSSION TITLE
Verify Mobile Number / Verify Email functions working?

### DISCUSSION THREAD

**Discussion by t-anjan** [tanmoysrt/pyaadhaar#21] [Q&A]
Hello,

I have been trying for a couple of days now, but I am unable to get the hash of mobile number or hash of email to match with the hash extracted from the QR. Basically, the `verifyMobileNumber` and `verifyEmail` functions always give false. I have tried with multiple Aadhaar cards.

@tanmoysrt - has it ever worked for you? Any ideas?

**Comment by tanmoysrt** [comment]
> Hello,
> 
> I have been trying for a couple of days now, but I am unable to get the hash of mobile number or hash of email to match with the hash extracted from the QR. Basically, the `verifyMobileNumber` and `verifyEmail` functions always give false. I have tried with multiple Aadhaar cards.
> 
> @tanmoysrt - has it ever worked for you? Any ideas?

Before 1 year, it was working. Need to check again. Little bit busy in work.Whenever I will be free, I will try to go through all the functionality of the library again.

**Reply by t-anjan** [reply]
I debugged the code. The hash of the input mobile and email get generated and seem correct. But the hash being read from the QR data is completely different. I am not sure if the bytearray indexes for the hashes are being read correctly.

```

    # Return hash of the email id.
    # Check the value of Email_mobile_present_bit_indicator_value:
    #   * If its 3 then first read mobile from index (Byte array length - 1 - 256) and
    #       then email from index (Byte array length - 1 - 256 - 32) in reverse order. Each value will be of fix size of 32 byte.
    #
    #   * If Email_mobile_present_bit_indicator_value is 1 then only mobile is present.
    #
    #   * If Email_mobile_present_bit_indicator_value is 2 then only email is present.
    #
    #   * If Email_mobile_present_bit_indicator_value is 0 then no mobile or email present.
    #
    # Email and Mobile value will available in byte. Convert into Hexadecimal String
    def sha256hashOfEMail(self) -> str:
        tmp = ""

        if self.isEmailRegistered() and self.isMobileNoRegistered():
            start_idx = len(self.decompressed_array)-256-32-32
            end_idx = len(self.decompressed_array)-256-32
            tmp = self.decompressed_array[start_idx : end_idx].hex()
        elif self.isEmailRegistered() and not self.isMobileNoRegistered():
            start_idx = len(self.decompressed_array)-256-32
            end_idx = len(self.decompressed_array)-256
            tmp = self.decompressed_array[start_idx : end_idx].hex()

        return tmp

    # Return hash of the mobile number
    def sha256hashOfMobileNumber(self) -> str:
        tmp = ""

        if self.isMobileNoRegistered():
            start_idx = len(self.decompressed_array)-256-32
            end_idx = len(self.decompressed_array)-256
            tmp = self.decompressed_array[start_idx : end_idx].hex()

        return tmp
```

**Reply by tanmoysrt** [reply]
Do you have secure qr docs ?
old link is not opening

**Reply by t-anjan** [reply]
Yeah, the QR codes I am using are generated in 2023.

What do you mean by "old link"?

Edit: Now I understand that you were asking about the documentation of secure QR codes. I thought you were asking me if I had "secure" QR or the older "normal" QR printed on my Aadhaar card.

**Reply by tanmoysrt** [reply]
I mean there was an official docs from uidai back then, which described the cryptographic algo used in secure QR. Do you have any ?

**Reply by t-anjan** [reply]
This is the only doc which I am aware of: https://uidai.gov.in/images/resource/User_manulal_QR_Code_15032019.pdf

Also attaching the doc directly here.
[User_manulal_QR_Code_15032019.pdf](https://github.com/user-attachments/files/16508516/User_manulal_QR_Code_15032019.pdf)
